File: Iteration2/code/CSE3063GRP12/src/features/transcript/TranscriptController.java
package features.transcript;

import java.util.*;

import core.general_providers.SessionController;
import core.general_providers.TerminalManager;
import core.models.concretes.Course;
import core.models.concretes.Semester;
import core.models.concretes.Student;
import core.models.concretes.Transcript;
import core.repositories.CourseRepository;
import core.repositories.TranscriptRepository;
import features.main_menu.MenuController;

public class TranscriptController {
    private TranscriptView transcriptView;
    private TranscriptRepository transcriptRepository;
    private CourseRepository courseRepository;

    public TranscriptController() {
        this.transcriptView = new TranscriptView();
        this.transcriptRepository = new TranscriptRepository();
        this.courseRepository = new CourseRepository();
        handleTranscript();
    }

    private Transcript fetchTranscript(String studentId) {
        try {
            return transcriptRepository.getTranscript(studentId);
        } catch (Exception e) {
            return null;
        }

    }

    private String getUserInput() {
        return TerminalManager.getInstance().read();

    }

    private void navigateToMenu() {
        new MenuController();
    }

    private void handleTranscript() {
        try {
            SessionController sessionController = SessionController.getInstance();
            Transcript transcript = fetchTranscript(((Student) sessionController.getCurrentUser()).getUserName());

            Map<Integer, Semester> semesters = transcript.getListOfSemester();
            if (semesters == null) {
                System.out.println(
                        "Student doesn't have a semester");
                transcriptView.showQuitMessage();
                if (getUserInput().equals("q"))
                    navigateToMenu();

            } else {
                for (Semester semester : semesters.values()) {
                    System.out.println(
                            "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------");
                    ArrayList<String> courseIdList = new ArrayList<>(semester.getListOfCoursesTaken().keySet());
                    ArrayList<Course> courses = courseRepository.findCoursesWithCourseIds(courseIdList);

                    System.out.println("Course Code \t Course Name \t Course Credit \t Course Grade");
                    for (Course course : courses) {

                        System.out.println(course.getCourseCode() + "\t" + course.getName() + "\t" + course.getCredit()
                                + "\t" + semester.getListOfCoursesTaken().get(course.getCourseCode()));
                    }
                }
                System.out.println(
                        "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------");
                transcriptView.showQuitMessage();
                if (getUserInput().equals("q"))
                    navigateToMenu();
            }

        } catch (Exception e) {
            e.printStackTrace();
        }

    }
}

File: Iteration3/code/CSE3063GRP12/src/features/student/transcript/TranscriptController.py
from features.student.transcript.TranscriptView import TranscriptView
from core.repositories.TranscriptRepository import TranscriptRepository
from core.repositories.CourseRepository import CourseRepository
from core.general_providers.SessionController import SessionController
from core.exceptions.UnexpectedInputException import UnexpectedInputException

class TranscriptController:
    def __init__(self):
        self.transcript_view = TranscriptView()
        self.transcript_repository = TranscriptRepository()
        self.course_repository = CourseRepository()
        self.handle_transcript()
    
    def fetch_transcript(self, student_id):
        try:
            return self.transcript_repository.get_transcript(student_id)
        except Exception as e:
            print(f"An error occurred while fetching transcript: {e}")
            return None
    
    def navigate_to_menu(self):
        from features.main_menu.MenuController import MenuController
        MenuController()
     
    def handle_transcript(self):
        try:
            session_controller = SessionController.getInstance()
            transcript = self.fetch_transcript(session_controller.get_current_user().userName)

            semesters = transcript.listOfSemesters
            if semesters is None:
                input_message = input("Press q to return Main Menu: ")
                if input_message == "q":
                    self.navigate_to_menu()
                else:
                    raise UnexpectedInputException("Unexpected input!")
            else:
                print(">>--->TRANSCRIPT========================================================================================================================================================|||")
                print()

                for semester in semesters.values():
                    print("===========================================================================================================================================================================")
                    print("# -> Year ({}): --Semester [{}]--".format(semester.semesterNo // 2 if semester.semesterNo % 2 == 0 else (semester.semesterNo + 1) // 2, semester.semesterNo))
                    print("---------------------------------------------------------------------------------------------------------------------------------------------------------------------------")

                    course_id_list = list(semester.listOfCoursesTaken.keys())
                    courses = self.course_repository.findCoursesWithCourseIds(course_id_list)

                    print("\t{: <15}{: <80}{: <20}{: <20}".format("Course Code", "Course Name", "Course Credit", "Course Grade"))
                    print("---------------------------------------------------------------------------------------------------------------------------------------------------------------------------")
                    print()

                    for course in courses:
                        
                        course_grade = semester.get("listOfCoursesTaken").get(course.courseCode)
                        if course_grade is None:
                            course_grade = "NA"
                        
                        print("\t{: <15}{: <85}{: <20}{: <20}".format(course.courseCode, course.name, course.credit, semester.listOfCoursesTaken[course.courseCode]))

                    print()
                    print("===========================================================================================================================================================================")
                    print(">>---> Completed Credits: {}".format(semester.credits_completed))
                    print(">>---> Taken Credits: {}".format(semester.credits_taken))
                    print(">>---> Yano: {}".format(semester.yano))
                    print(">>---> Gano: {}".format(semester.gano))
                    print()

                print("|||========================================================================================================================================================TRANSCRIPT<---<<")
                print("###===>> Completed Credits: {}".format(transcript.totalCreditCompleted))
                print("###===>> Taken Credits: {}".format(transcript.totalCreditTaken))
                print("###===>> Gano: {}".format(transcript.gano))
                print("|||========================================================================================================================================================TRANSCRIPT<---<<")
                
                while True:
                    input_message = input("Press q to return Main Menu: ")
                    if input_message == "q":
                        self.navigate_to_menu()
                        break
                    else:
                        raise UnexpectedInputException("Unexpected input!")
        except Exception as e:
            print(e)
        
